pkg/expression/builtin_compare.go

```diff
@@ -141,6 +141,8 @@ func (c *coalesceFunctionClass) getFunction(ctx sessionctx.Context, args []Expre
 		return nil, err
 	}
 
+	bf.tp = resultFieldType
+
 	bf.tp.AddFlag(resultFieldType.GetFlag())
```

Comment:
This line can be removed

Reply:
And I found there is lots of code(L145-L197) to adjust `resultFieldType` according to input fieldtype, can you double check if we can use `resultFieldType, err := InferType4ControlFuncs(xxx)`, and remove these extra adjustments? Because theoretically, the type infer for `coalesce` should be the same as `case when/ifnull`, which all use `InferType4ControlFuncs`

Reply:
> And I found there is lots of code(L145-L197) to adjust `resultFieldType` according to input fieldtype, can you double check if we can use `resultFieldType, err := InferType4ControlFuncs(xxx)`, and remove these extra adjustments? Because theoretically, the type infer for `coalesce` should be the same as `case when/ifnull`, which all use `InferType4ControlFuncs`

After check, I think we can not remove these extra adjustments.